### DISCUSSION TITLE
Recommendation to best use resources per self-hosted runner?

### DISCUSSION THREAD

**Discussion by oerp-odoo** [community/community#46142] [Actions] [answered]
### Select Topic Area

Question

### Body

I am setting up self-hosted runners and I am thinking if there are any recommendations, rules of thumbs etc how many runners there should be per host/CPU count or something?

For example, if I have Server with 4 CPU, should I have 4 runners there to best use 4 CPU or maybe 1 runner per 2CPU?

Currently I have initial server with 4CPU and 2 runners there.

Also currently self-hosted runner seems to not run jobs in parallel if there is single runner, even if you have lots of CPUs, so to have parallelism, you need at least 2 runners. But I am not sure how effective each runner is, depending on available resources on single machine or per host?

**Comment by vitalykarasik** [comment, marked as answer]
"how many runners there should be per host/CPU count" : As usual - it depends on your workload. For instance, for Android ROM builds I use runners with 128 CPUs.

"How many runners per host?" : I prefer a single runner per host, i.e. I launch a cloud VM with ephemeral GHA runner for each relevant job. This makes orchestration easier.

**Reply by oerp-odoo** [reply]
So your 128CPUs are split between many (temp) hosts I assume? Currently my workload is not big (as I am just building new infrastructure and testing this), but when many developers start pushing code, there might be a need for lots of parallel jobs to run.

**Reply by vitalykarasik** [reply]
No - I have a few CPU-intensive jobs, and for these jobs I'm launching multi-CPUs runners, runner per job. 

But your case may be different. If you don't need many CPUs for your job, you can launch 1 or 2 CPUs instances. By "it depends" I meant that sizing is 100% dependent on your workflow, exactly as for backend applications. 
Regarding 'how many runners per VM?' question - I'm using cloud VMs, so I prefer single runner per host, which makes orchestration simple (IMHO). If you have on-premises servers, multiple runners per host make sense.
